### DISCUSSION TITLE
Can sass-loader understand package.json "exports" field?

### DISCUSSION THREAD

**Discussion by krutoo** [webpack/webpack#17990] [Q&A] [answered]
Hi, in my UI library i have some scss files. This library build from TS and SCSS files:

- JS files compiles from TS files from src folder to build folder
- CSS files just copies from src folder to build folder

In package.json of my library i have:
```json
{
  "name": "@my-org/my-library",
  "exports": {
    "./utils.scss": "./build/utils.scss"
  },
}
```

In the project in which I use my library I tried imports:
```scss
@use "@my-org/my-library/utils";

// or
@use "@my-org/my-library/utils.scss";

// or
@use "node_modules/@my-org/my-library/utils";

// or
@use "node_modules/@my-org/my-library/utils.scss";
```

But it is not working, in this project i use sass-loader with rspack but it crashes with error:
```
Module build failed:
  ╰─▶   × Can't find stylesheet to import.
        │   ╷
        │ 1 │ @use '@my-org/my-library/colors';
```

Can sass-loader understand package.json "exports" field?

**Comment by alexander-akait** [comment]
Did you try it using webpack? Because yes, it should works

**Reply by alexander-akait** [reply]
We have test cases https://github.com/webpack-contrib/sass-loader/blob/master/test/node_modules/package-with-exports-and-custom-condition/package.json

**Reply by krutoo** [reply]
@alexander-akait i tried with webpack and it is not working:
```
Module build failed (from ./node_modules/sass-loader/dist/cjs.js):
Can't find stylesheet to import.
  ╷
1 │ @use 'node_modules/@my-org/my-library/utils';
  │ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ╵
```

**Reply by alexander-akait** [reply]
What is

> @use "@my-library/utils"

?

Sounds like you have `@my-library` org and `utils` package with `exports`. It means you should use `@my-library/utils/utils.scss`.

Also you can have `default` export:
```
"exports": {
    ".": "./build/utils.scss"
},
```

**Reply by krutoo** [reply]
@alexander-akait I apologize for the misleading, i updated code snippets

**Reply by krutoo** [reply]
@alexander-akait I create repo with demo:
https://github.com/krutoo/sass-loader-package-exports

Just make `npm i` and `npm run dev` in root folder

It doesn't matter whether there is an organization or not in package with scss exports

**Reply by alexander-akait** [reply, marked as answer]
Sorry for delay, there is an intresting problem... `exports` works fine, you have:
```scss
// @use "my-library/utils.scss"; // also not working
```

But it works fine, just uncomment it and check out.

Because `import * as mod from "something.js"` requires an extension, you need to write `scss` extension too for sass, i.e. `@use "my-library/utils.scss";`, theoretically you can make a small plugin that allow to use it without the `scss` extensions, but I don't feel it is a right solution, `import ... from "something.js"` and the `exports` field were originally designed to have an extension at the end and I have strong feel if sass wants to be align with javascript and don't create potential misunderstanding for developers we need to solve it or just improve sass documentation that recommends using the extension at the end.

```sass
@use "node_modules/my-library/utils.scss"; // not working
@use "node_modules/my-library/utils"; // also not working
```

Will not work, because the `exports` logic doesn't work in such case, you can test the same using JS

**test.mjs**

```
// import foo1 from "./node_modules/my-library/foo.js"; doesn't work
import foo2 from "my-library/foo.js"; // work
```

```json
{
  "name": "my-library",
  "version": "0.0.0",
  "description": "",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "exports": {
    "./utils.scss": "./src/utils.scss",
    "./foo.js": "./src/foo.js"
  }
}
```


@nex3 There's an interesting problem here, you might need to create RFC/discussion

**Reply by nex3** [reply]
This is handled by [the package importer proposal](https://github.com/sass/sass/blob/main/proposal/package-importer.d.ts.md), which looks for files that match `exports` using the same logic it looks for files on the filesystem (and so can handle resolving extensions, partials, and index files).

**Reply by krutoo** [reply]
@nex @alexander-akait This may be the wrong place to ask, but do you have any thoughts on how to properly define paths for the @use rule. 

Currently we are using paths like `@use "node_modules/@my-org/my-lib/utils"` and skip extension because this was once a rule in the **stylelint-config-sass-guidelines** package

As i understand now we should definitely use the extension when scss file is provided by exports field

**Reply by alexander-akait** [reply]
> As i understand now we should definitely use the extension when scss file is provided by exports field

I would say that this is a recommendation, you still may not specify this, but in order to be compatible with the `exports` field (and other JS logic) it is better to do it. Feel free to open an issue in `stylelint-config-sass-guidelines` and link this issue

**Reply by nex3** [reply]
The recommendation as soon as https://github.com/sass/sass/issues/2739 is released (in the next few days) will be to write `@use "pkg:@my-org/my-lib/utils"`.

**Reply by elrouss** [reply]
Nice solution. Thanks!
